### PR TITLE
Fix locality load balancing

### PR DESCRIPTION
**Pull request by howardjohn**
This PR includes 3 things:
* Fix for https://github.com/istio/istio/issues/20077. When there is no
locality, traffic will never be sent to the pod, even with locality lb
disabled. The fix here is setting locality to {} instead of nil
* Destinationrule cannot override mesh config to disable locality lb
* Fix https://github.com/istio/istio/issues/13317 by adding the test

The test added also tests the first two cases

Test is still failing locally so need to make a few more changes, WIP at the moment

### CODE REVIEW

**Comment by howardjohn** [comment]
fixes https://github.com/istio/istio/issues/18903

**Comment by howardjohn** [comment]
@hzxuzhonghu I think this is ready now

File: pilot/pkg/proxy/envoy/v2/eds.go [hunk 1]
```diff
@@ -687,19 +687,12 @@ func (s *DiscoveryServer) generateEndpoints(
 	}
 
 	// If locality aware routing is enabled, prioritize endpoints or set their lb weight.
-	if push.Mesh.LocalityLbSetting != nil {
-		// Make a shallow copy of the cla as we are mutating the endpoints with priorities/weights relative to the calling proxy
-		clonedCLA := util.CloneClusterLoadAssignment(l)
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/eds.go]: Why do you remove this

File: tests/integration/pilot/locality/distribute_test.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
+// Copyright 2019 Istio Authors
```
Comment (hzxuzhonghu) [inline tests/integration/pilot/locality/distribute_test.go]: 2020

File: pilot/pkg/proxy/envoy/v2/eds.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/eds.go]: Good catch, this as accidental. I am surprised tests pass...

Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/eds.go]: I added it back

**Pull request merged**